bucket/drivers/s3/s3.go

```diff
@@ -0,0 +1,113 @@
+package s3
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/sirupsen/logrus"
+)
+
+type S3Driver struct {
+	//svc is IAM S3 service client
+	svc    *s3.S3
+	config *aws.Config
+}
+
+func New(config *aws.Config) *S3Driver {
+	sess, _ := session.NewSession(config)
+	// Create S3 service client
+	svc := s3.New(sess)
+	return &S3Driver{
+		svc:    svc,
+		config: config,
+	}
+}
+
+func (d *S3Driver) CreateBucket(name string) (string, error) {
+	_, err := d.svc.CreateBucket(&s3.CreateBucketInput{
+		Bucket: aws.String(name),
+		CreateBucketConfiguration: &s3.CreateBucketConfiguration{
+			LocationConstraint: aws.String(*d.config.Region),
+		},
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				logrus.Error("Bucket name already in use!")
+				panic(err)
```

Comment:
If BucketAlreadyExists, let's make this call Idempotent and return the existing bucket ID/name without an error.

This makes it easier for the controller to know when a bucket already exists or has been created.

Reply:
Did’t intend to “panic” here. It is a residue of some testing code that I had added and forgot to remove. Will remove it.
1. BucketAlreadyExists =>  Bucket name needs to be updated since it is not globally unique.  return “”, err
2. BucketAlreadyOwnedByYou => Bucket was created by the current requester. =>   return id, nil
3. Any Other Error => return “”, err
4. Success => return id, nil
So we can combine (1, 3 )   and (2, 4). Does that make sense?
But note that if the controller needs to know if the bucket was created with the current call vs if it already existed, then combining 2, 4 will not work.

Reply:
Yes exactly, that makes sense.

The controller will not care about whether the bucket already exists. In CSI, use the concept of idempotency a lot to ensure that if multiple requests reach the driver, the driver will respond the same way. 

Here is an example in CSI where we return early if the desired expansion size equals the current size already:
https://github.com/libopenstorage/openstorage/blob/master/csi/controller.go#L711-L727

Another example is in sdk.DeleteVolume, we return success if the volume cannot be found:
https://github.com/libopenstorage/openstorage/blob/master/api/server/sdk/volume_ops.go#L336-L347

This is effectively saying: you requested a volume to be deleted. It has already been deleted, so the call is successful.